### PR TITLE
For preview changes

### PR DESCRIPTION
**Pull request by daradurvs**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/MavenUtils.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    private static Process exec(String cmd) throws IOException, InterruptedException {
+        ProcessBuilder pb = new ProcessBuilder();
+        pb.redirectErrorStream(true);
+        pb.command(U.isWindows() ?
+            new String[] {"cmd", "/c", cmd} :
+            new String[] {cmd});
+
+        Process p = pb.start();
+
+        assert p.waitFor() == 0 : "Command=" + cmd + " couldn't be executed: "
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/MavenUtils.java]: 1. Can't assert some useful method.
```
int result = p.waitFor();
assert result == 0;
```
2. waitFor() can lead to thread hangs forever.
I think we have to execute command with timeout:
```
Timer timer = new Timer(true);
Process p = null;
            try {
                p = pb.start();
                timer.schedule(new KillProcessTimerTask(process), timeout);
                process.waitFor();
            } catch (InterruptedException e) {
                if (process != null) {
                    process.destroy();
                }
            } finally {
                timer.cancel();
            }

public class KillProcessTimerTask extends TimerTask {
    private Process p;

    public KillProcessTimerTask(Process p) {
        this.p = p;
    }

    public void run() {
        p.destroy();
    }
}

```

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java [hunk 2]
```diff
@@ -864,6 +867,105 @@ protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg, G
...
+     * @param clos IgniteClosure for post-configuration.
+     * @param jvmArgs Additional JVM arguments.
+     * @param resetDiscovery Reset DiscoverySpi.
+     * @return Started grid.
+     * @throws Exception If failed.
+     */
+    protected Ignite startGrid(String igniteInstanceName, String ver, IgniteConfiguration cfg,
+        IgniteInClosure<IgniteConfiguration> clos, Collection<String> jvmArgs,
+        boolean resetDiscovery) throws Exception {
+
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: Does empty line required?)

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java [hunk 3]
```diff
@@ -864,6 +867,105 @@ protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg, G
...
+        assert !isFirstGrid(igniteInstanceName);
+
+        if (cfg == null)
+            cfg = optimize(getConfiguration(igniteInstanceName));
+
+        Collection<String> filteredJvmArgs = new ArrayList<>();
+
+        filteredJvmArgs.add("-ea");
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: I think we don't want to spread logic of modifying jvm arguments from IgniteProcessProxy.
In my opinion we have to make method `filteredJvmArgs` in IgniteProcessProxy and override it if necessarry.

I implement it in my unmerged commit:
https://github.com/nizhikov/ignite/pull/2/commits/1857a95185266e256b269897e577cae18db7a577#diff-74656e7b5fd29f2fccd75f89febeb052R81

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java [hunk 4]
```diff
@@ -864,6 +867,105 @@ protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg, G
...
+        }
+
+        String classPath = System.getProperty("java.class.path");
+
+        String[] paths = classPath.split(File.pathSeparator);
+
+        StringBuilder pathBuilder = new StringBuilder();
+
+        for (String path : paths) {
+            if (path.contains("ignite\\modules\\core\\target\\classes"))
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: We need to use `File.separator` to separate directories.
`\\` is Windows specific

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/TestMultiVersionMode.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    @Override protected boolean isMultiJvm() {
+        return true;
+    }
+
+    /** */
+    public void testJoinMultiVersionTopology() throws Exception {
+        try {
+            startGrid(0);
+            
+            startGrid("testMultiVersion", "2.1.0", null);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/TestMultiVersionMode.java]: What happen when version become 2.2.0 or 2.1.1?
I think current version should be used from pom.xml somehow.
May be we can get it from System properties or Plugin properties?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java [hunk 6]
```diff
@@ -69,7 +75,12 @@ public static void main(String[] args) throws Exception {
 
         X.println("Starting Ignite Node... Args=" + Arrays.toString(args));
 
-        IgniteConfiguration cfg = readCfgFromFileAndDeleteFile(args[0]);
+        IgniteConfiguration cfg;
+
+        if (args.length > 1)
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: Do we need to check args != null?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java [hunk 7]
```diff
@@ -69,7 +75,12 @@ public static void main(String[] args) throws Exception {
...
+            cfg = readCfgFromFileAndDeleteFile(args[0], args[1]);
+        else
+            cfg = readCfgFromFileAndDeleteFile(args[0]);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: Do we need to check args.length != 0?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java [hunk 8]
```diff
@@ -135,18 +146,69 @@ public static void storeToFile(IgniteConfiguration cfg, String fileName,
     }
 
     /**
+     * Stores {@link IgniteInClosure} to file as xml.
+     *
+     * @param clos IgniteClosure.
+     * @param cfgFileName A name of file where the configuration was stored.
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: What do we store? IgniteInClousre or configuration?
I don't get it from javadoc

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java [hunk 9]
```diff
@@ -135,18 +146,69 @@ public static void storeToFile(IgniteConfiguration cfg, String fileName,
...
+     * @param cfgFileName Configuration file name.
+     * @param closFileName Closure file name.
+     * @return Read and prepared configuration.
+     * @throws IOException If failed.
+     * @throws IgniteCheckedException On error.
+     * @see #storeToFile(IgniteConfiguration, boolean)
+     * @see #storeToFile(IgniteInClosure, String)
+     */
+    @SuppressWarnings("unchecked")
+    private static IgniteConfiguration readCfgFromFileAndDeleteFile(String cfgFileName, String closFileName)
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: Second word file is uneccessary

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java [hunk 10]
```diff
@@ -135,18 +146,69 @@ public static void storeToFile(IgniteConfiguration cfg, String fileName,
...
     /**
+     * Stores {@link IgniteInClosure} to file as xml.
+     *
+     * @param clos IgniteClosure.
+     * @param cfgFileName A name of file where the configuration was stored.
+     * @return A name of file where the given closure was stored.
+     * @throws IOException If failed.
+     */
+    public static String storeToFile(IgniteInClosure clos, String cfgFileName) throws IOException {
+        String fileName = cfgFileName + IGNITE_CONFIGURATION_CLOSURE_POSTFIX;
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: CONFIGURATION -> CFG
CLOSURE -> CLR

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java (hunk 4 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: Good catch on the bag! Thanks!

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/TestMultiVersionMode.java (hunk 5 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/TestMultiVersionMode.java]: Not sure. Will be discussed on the dev-list.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java (hunk 6 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: It may have length 0 but will not be null in case of normal execution.
But I can't find such information in JLS: http://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java (hunk 7 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: The check weren't here before.
But I agree, it needs. Added.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java (hunk 9 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: Method was renamed to be more clear.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java (hunk 8 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: It was refactored to be more predictable. Thanks!

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java (hunk 10 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteNodeRunner.java]: Variable was removed during refactoring.

File: modules/core/src/test/java/org/apache/ignite/MavenUtils.java (hunk 1 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/MavenUtils.java]: 1. Agree, fixed.
2. In case of downloading file via maven it may take significant time, therefore I don't want specify timeout. But this class will be used only in tests and it will be managed by "tests execution timeout".

What do you think about 2?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java (hunk 2 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/MavenUtils.java (hunk 1 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/MavenUtils.java]: I think that I'm tired of timeouted Ignite suite on TC already :)
Do we need one more possible way for timeout?
Can we just set timeout to 10 minutes or similar?

Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/MavenUtils.java]: Ok, it will be remade.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java (hunk 3 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: Constructor was extended:
https://github.com/daradurvs/ignite/pull/17/files#diff-dbed5523c9851122582415113178aa0aR152
I need an opportunity to pass custom arguments.
Anonymous class with method overriding will complicate the code.
I think constructor extension more clear than anonymous class creation. Don't you?

Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: It is matter of taste. I don't argue if you leave it unchanged.

My point is - we have some logic for modifying class path. 
It is function with signature `List<String> -> List<String>`. 
Constructing `List<String>` for source classpath is responsibility of `IgniteProcessProxy`. 
It knows where to get all classpath arguments. 
If someone change logic `how construct classpath` in future one has to modify all places where constructor argument passed to process proxy.

In short words it DRY principle.

Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java]: Thanks for explanation. 
I think you are right.
It will be refactored.

File: modules/core/src/test/java/org/apache/ignite/MavenUtils.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        ProcessBuilder pb = new ProcessBuilder();
+        pb.redirectErrorStream(true);
+
+        pb.command(U.isWindows() ?
+            new String[] {"cmd", "/c", cmd} :
+            new String[] {"/bin/bash", "-c", cmd});
+
+        final Process p = pb.start();
+
+        ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/MavenUtils.java:151]: Variable can be omitted.
`Executors.newSingleThreadExecutor().submit(...);`

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java [hunk 12]
```diff
@@ -864,6 +867,111 @@ protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg, G
...
+                        filteredJvmArgs.add(arg);
+                }
+
+                String classPath = System.getProperty("java.class.path");
+
+                String[] paths = classPath.split(File.pathSeparator);
+
+                StringBuilder pathBuilder = new StringBuilder();
+
+                String corePathTemplate = "ignite.modules.core.target.classes".replace(".", File.separator);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java:948]: After `replace` it not a template but a corePath itself. Isn't it?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java [hunk 13]
```diff
@@ -864,6 +867,111 @@ protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg, G
...
+        IgniteInClosure<IgniteConfiguration> clos, final Collection<String> jvmArgs,
+        boolean resetDiscovery) throws Exception {
+        assert isMultiJvm() : "MultiJvm mode must be switched on for the node stop properly.";
+
+        assert !isFirstGrid(igniteInstanceName) : "Please, start node of current version first.";
+
+        if (cfg == null)
+            cfg = optimize(getConfiguration(igniteInstanceName));
+
+        final String pathToArtifact = MavenUtils.getPathToIgniteCoreArtifact(ver);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java:929]: Is this variable necessarry?
It has to be declared closer to usage(first usage on line 955)

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java (hunk 12 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java:948]: Not sure.
It isn't full core path, just a part of it.

File: modules/core/src/test/java/org/apache/ignite/MavenUtils.java (hunk 11 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/MavenUtils.java:151]: Will be fixed.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java (hunk 13 above)
Comment (daradurvs) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/GridAbstractTest.java:929]: Will be fixed.

**Comment by daradurvs** [comment]
see #19
